usr/sbin/p11sak/p11sak.c

```diff
@@ -1910,104 +2051,264 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
         }
     }
 
-    ret = funcs->C_FindObjectsFinal(session);
-    if (ret != CKR_OK) {
-        printf("C_FindObjectsFinal failed (error code 0x%lX)\n", ret);
+    rc = funcs->C_FindObjectsFinal(session);
+    if (rc != CKR_OK) {
+        printf("C_FindObjectsFinal failed (error code 0x%lX: %s)\n", rc,
+                p11_get_ckr(rc));
         goto done;
     }
 
-    ret = CKR_OK;
+    rc = CKR_OK;
 
     done: free(label);
     free(keytype);
-    return ret;
+    return rc;
 }
 
-static CK_RV execute_cmd(CK_SESSION_HANDLE session, CK_SLOT_ID slot, p11sak_cmd cmd,
-        p11sak_kt kt, CK_ULONG keylength, CK_ULONG exponent, char *ECcurve,
-        char *label, char *attr_string, int long_print) {
-    CK_RV ret;
+/**
+ * Delete objects
+ */
+static CK_RV deleteKey(CK_SESSION_HANDLE session, p11sak_kt kt, char *rm_label,
+        CK_BBOOL *forceAll) {
+    CK_ULONG keylength, count;
+    CK_OBJECT_CLASS keyclass;
+    CK_OBJECT_HANDLE hkey;
+    char *keytype = NULL;
+    char *label = NULL;
+    CK_RV rc;
+
+    int nread;
+    char *buff = NULL;
+    size_t buflen;
+
+    rc = tok_key_list_init(session, kt, label);
+    if (rc != CKR_OK) {
+        printf("Init token key list failed (error code 0x%lX: %s)\n", rc,
+                p11_get_ckr(rc));
+        return rc;
+    }
+
+    while (1) {
+        rc = funcs->C_FindObjects(session, &hkey, 1, &count);
```

Comment:
You only get 1 find result per loop iteration. It works that way, but it would perform much better if you get multiple (say 10 or 20) results with one call to C_FindObjects and the have a secondary loop to iterate over the results (0 .. count -1).

Reply:
This will become an item for future implementation.